src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,317 @@
+        PooledObject<SerializableBytes.ReadWriteStream> GetTempStream(out Stream stream)
         {
-            using var reader = ObjectReader.GetReader(stream, leaveOpen: true, cancellationToken);
+            var pooledObject = s_streamPool.GetPooledObject();
+            var tempStream = pooledObject.Object;
+            tempStream.Position = 0;
 
-            // Ensure that no invariants were broken and that both sides of the communication channel are talking about
-            // the same pinned solution.
-            var responseSolutionChecksum = Checksum.ReadFrom(reader);
-            Contract.ThrowIfFalse(solutionChecksum == responseSolutionChecksum);
+            // Don't truncate the stream as we're going to be writing to it multiple times.  This will allow us to
+            // reuse the internal chunks of the buffer, without having to reallocate them over and over again.
+            // Note: this stream internally keeps a list of byte[]s that it writes to.  Each byte[] is less than the
+            // LOH size, so there's no concern about LOH fragmentation here.
+            tempStream.SetLength(0, truncate: false);
+            stream = tempStream;
+            return pooledObject;
+        }
+    }
+
+    public static ValueTask ReadDataAsync<T, TArg>(
+        PipeReader pipeReader, int objectCount, ISerializerService serializerService, Action<Checksum, T, TArg> callback, TArg arg, CancellationToken cancellationToken)
+    {
+        // Suppress ExecutionContext flow for asynchronous operations operate on the pipe. In addition to avoiding
+        // ExecutionContext allocations, this clears the LogicalCallContext and avoids the need to clone data set by
+        // CallContext.LogicalSetData at each yielding await in the task tree.
+        //
+        // ⚠ DO NOT AWAIT INSIDE THE USING. The Dispose method that restores ExecutionContext flow must run on the
+        // same thread where SuppressFlow was originally run.
+        using var _ = FlowControlHelper.TrySuppressFlow();
+        return ReadDataSuppressedFlowAsync(pipeReader, objectCount, serializerService, callback, arg, cancellationToken);
 
-            for (int i = 0, n = objectCount; i < n; i++)
+        static async ValueTask ReadDataSuppressedFlowAsync(
+            PipeReader pipeReader, int objectCount, ISerializerService serializerService, Action<Checksum, T, TArg> callback, TArg arg, CancellationToken cancellationToken)
+        {
+            using var pipeReaderStream = pipeReader.AsStream(leaveOpen: true);
+
+            for (var i = 0; i < objectCount; i++)
             {
+                // First, read the sentinel byte and the length of the data chunk we'll be reading.
+                const int HeaderSize = sizeof(byte) + sizeof(int);
+                var lengthReadResult = await pipeReader.ReadAtLeastAsync(HeaderSize, cancellationToken).ConfigureAwait(false);
+                if (lengthReadResult.IsCanceled || lengthReadResult.IsCompleted)
```

Comment:
>IsCompleted [](http://example.com/codeflow?start=68&length=11)

This is unintuitive (to me). Can you explain a bit?

Reply:
removing this.

Reply:
had a misunderstanding about the api.  this is not needed.